py/desispec/io/spectra.py

```diff
@@ -0,0 +1,285 @@
+    all_hdus.append(fits.PrimaryHDU(header=hdr))
+
+    # Next is the fibermap
+    fmap = spec.fmap.copy()
+    fmap.meta["EXTNAME"] = "FIBERMAP"
+    hdu = fits.convenience.table_to_hdu(fmap)
+
+    # Add comments for fibermap columns.
+    comments = spectra_comments()
+    num_columns = len(comments)
+    for i in range(1, 1 + num_columns):
+        key = "TTYPE{}".format(i)
+        name = hdu.header[key]
+        comment = comments[name]
+        hdu.header[key] = (name, comment)
+
+    all_hdus.append(hdu)
+
+    # Now append the data for all bands
+
+    for band in spec.bands:
+        hdu = fits.ImageHDU(name="{}_WAVELENGTH".format(band.upper()))
+        hdu.header["BUNIT"] = "Angstrom"
+        hdu.data = spec.wave[band].astype("f4")
+        all_hdus.append(hdu)
+
+        hdu = fits.ImageHDU(name="{}_FLUX".format(band.upper()))
+        if units is None:
+            hdu.header["BUNIT"] = "1e-17 erg/(s cm2 Angstrom)"
+        else:
+            hdu.header["BUNIT"] = units
+        hdu.data = spec.flux[band].astype("f4")
+        all_hdus.append(hdu)
+
+        hdu = fits.ImageHDU(name="{}_IVAR".format(band.upper()))
+        hdu.data = spec.ivar[band].astype("f4")
+        all_hdus.append(hdu)
+
+        if spec.mask is not None:
+            hdu = fits.CompImageHDU(name="{}_MASK".format(band.upper()))
+            hdu.data = spec.mask[band].astype(np.uint32)
+            all_hdus.append(hdu)
+
+        if spec.resolution_data is not None:
+            hdu = fits.ImageHDU(name="{}_RESOLUTION".format(band.upper()))
+            hdu.data = spec.resolution_data[band].astype("f4")
+            all_hdus.append(hdu)
+
+        if spec.extra is not None:
+            for ex in spec.extra[band].items():
+                hdu = fits.ImageHDU(name="{}_{}".format(band.upper(), ex[0]))
+                hdu.data = ex[1].astype("f4")
+                all_hdus.append(hdu)
+
+    all_hdus.writeto("{}.tmp".format(outfile), overwrite=True, checksum=True)
+    os.rename("{}.tmp".format(outfile), outfile)
+
+    return outfile
+
+
+def read_spectra(infile, single=False):
+    """
+    Read Spectra object from FITS file.
+
+    This reads data written by the write_spectra function.  A new Spectra
+    object is instantiated and returned.
+
+    Args:
+        infile (str): path to read
+        single (bool): if True, keep spectra as single precision in memory.
+
+    Returns (Spectra):
+        The object containing the data read from disk.
+
+    """
+
+    ftype = np.float64
+    if single:
+        ftype = np.float32
+
+    infile = os.path.abspath(infile)
+    if not os.path.isfile(infile):
+        raise IOError("{} is not a file".format(infile))
+
+    hdus = fits.open(infile, mode="readonly")
+    nhdu = len(hdus)
+
+    # load the metadata.  FITS will capitalize all our metadata
+    # which is annoying.  Here we lower it all.
+
+    meta = {}
+    for k, v in hdus[0].header.items():
+        if k != "EXTNAME":
+            meta[k.lower()] = v
```

Comment:
Elsewhere we leave these as upper case.  Let's just cast this to a dictionary via `meta = dict(hdus[0].header)` and leave them as is.  Or you could even leave it as a `astropy.io.fits.header.Header` object in which case either upper or lowercase keys will work.  Header object are much less efficient to access, but that only matters if we are looking up keywords many many times.